src/celeritas/phys/RootPrimaryGenerator.cc

```diff
@@ -0,0 +1,121 @@
+//! \file celeritas/phys/RootPrimaryGenerator.cc
+//---------------------------------------------------------------------------//
+#include "RootPrimaryGenerator.hh"
+
+#include <TFile.h>
+#include <TLeaf.h>
+#include <TTree.h>
+
+#include "corecel/Macros.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/io/Logger.hh"
+#include "celeritas/ext/ScopedRootErrorHandler.hh"
+#include "celeritas/phys/ParticleParams.hh"
+#include "celeritas/phys/Primary.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from input file, number of events, and primaries to be sampled per
+ * event.
+ */
+RootPrimaryGenerator::RootPrimaryGenerator(std::string const& filename,
+                                           SPConstParticles params,
+                                           size_type num_events,
+                                           size_type primaries_per_event,
+                                           unsigned int seed)
+    : params_(std::move(params))
+    , num_events_{num_events}
+    , primaries_per_event_(primaries_per_event)
+{
+    CELER_EXPECT(!filename.empty());
+    CELER_VALIDATE(num_events > 0 && primaries_per_event > 0,
+                   << "'num_events' and 'primaries_per_event options in input "
+                      "file must be set and positive");
+
+    ScopedRootErrorHandler scoped_root_error;
+    tfile_.reset(TFile::Open(filename.c_str(), "read"));
+    CELER_ENSURE(tfile_->IsOpen());
+
+    ttree_.reset(tfile_->Get<TTree>("primaries"));
+    CELER_ENSURE(ttree_);
+    CELER_VALIDATE(ttree_->GetEntries() > 0,
+                   << "TTree `primaries` in '" << tfile_->GetName()
+                   << "' has zero entries");
+    scoped_root_error.throw_if_errors();
+
+    entry_selector_ = std::uniform_int_distribution<>(0, primaries_per_event);
+    rng_.seed(seed);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Populate a vector of primaries from the ROOT offloaded primaries input file
+ * by uniformly selecting random primaries throughout the file.
+ */
+auto RootPrimaryGenerator::operator()() -> result_type
+{
+    if (event_count_.get() == num_events_)
+    {
+        return {};
+    }
+
+    ScopedRootErrorHandler scoped_root_error;
+
+    // Fetch ROOT leaves
+    // TODO: Make these private members?
+    auto lpid = ttree_->GetLeaf("particle");
+    CELER_ASSERT(lpid);
+    auto lenergy = ttree_->GetLeaf("energy");
+    CELER_ASSERT(lenergy);
+    auto lpos = ttree_->GetLeaf("pos");
+    CELER_ASSERT(lpos);
+    auto ldir = ttree_->GetLeaf("dir");
+    CELER_ASSERT(ldir);
+    auto ltime = ttree_->GetLeaf("time");
+    CELER_ASSERT(ltime);
+
+    TrackId track_id{0};
+    result_type result(primaries_per_event_);
+    for (auto i : range(primaries_per_event_))
+    {
+        ttree_->GetEntry(entry_selector_(rng_));
```

Comment:
I don't think we should be sampling individual primaries from the file since the primaries are highly correlated inside the event. I was imagining we'd sample *events* from the input, with a user parameter describing how many "ROOT file events" per "RPG event". This way we could run individual pions for the test beam, and then sample 700 different showers for a single "14 TeV" event.

Perhaps @amandalund has an opinion?

Reply:
Yeah, that would definitely be a more realistic input. We might even be able to do that in the `RootEventReader` (optionally return a random subset of events)?

Reply:
To be honest I'd forgotten about RootEventReader 😞 which could perhaps have saved some work here...